### PR TITLE
feat: add dependency name for package.json detected errors

### PR DESCRIPTION
**Pull request by barakyosi**
## Summary
Update:
This PR adds package name to issues related to a `package.json` file.

image

Note - this PR only handles diagnostics that are added in a regular flow. Thrown diagnostics are currently skipped.
As license check uses the throw diagnostic flow, this PR no longer fix the issue it was open to fix 😞 

Adding this message to thrown diagnostics would require more effort, and currently not part of this PR 😅 

Some thoughts how to do so:
1. Use synced diagnostic flow in [`normalizeLicense`](https://github.com/romefrontend/rome/blob/5c9f1898d2fc87d6ff8034eb960aa89068309e80/internal/codec-js-manifest/index.ts#L216).
2. Add package name in [caught diagnostics](https://github.com/romefrontend/rome/blob/main/internal/core/server/fs/MemoryFileSystem.ts#L631).

That's my first PR here, would be happy for any feedback 😅 

## Test Plan

I tested it manually:
```
mkdir test-rome
npm init

Rename `dependencies` to `Dependencies` in package.json.

../rome/rome check
```

### CODE REVIEW

**Comment by sebmck** [comment]
I'm worried about the precedent this sets, since we're duplicating information that's already available in the filename in the error message.

Generally, this sort of information should be appended as diagnostic advice if we really want it.

That can be done in [`MemoryFileSystem`](https://github.com/romefrontend/rome/blob/main/internal/core/server/fs/MemoryFileSystem.ts#L652-L655) with something like:

```
import {manifestNameToString} from "@internal/codec-js-manifest";

normalizedDiagnostics.map((diag) => ({
    ..diag,
    description: {
        ...diag.description,
        advice: [
            ...diag.description.advice,
            {
                type: "log",
                category: "info",
                text: markup`Error occurred for package <emphasis>${manifestNameToString(manifest.name)}</emphasis> at <emphasis>${path.getParent()}</emphasis>`,
            },
        ],
    },
}));
```

This way, it will be present for all manifest diagnostics, regardless of where they come from.

What do you think?

>  Another option to solve this issue is to add public domain (or make the test case insensitive):

Oh yeah! This is actually done in #1030 but has yet to be merged (I'm going through our backlog now).

**Comment by barakyosi** [comment]
> Oh yeah! This is actually done in #1030 but has yet to be merged (I'm going through our backlog now).

That's great!
I believe this solves the root cause of the issue 😅 

> I'm worried about the precedent this sets, since we're duplicating information that's already available in the filename in the error message.

Yeah the package name is already a part of the filename and IMO it's pretty easy to find that out. However - as said in the issue, it might be useful for beginners to understand the error is not part of their code.
I'm good with closing this PR, but we can also continue with your suggestion if you'd like.

I had some troubles understanding where exactly to add this diagnostics normalization, as there's 2 places that add diagnostics in `MemoryFileSystem` - [1](https://github.com/romefrontend/rome/blob/main/internal/core/server/fs/MemoryFileSystem.ts#L659) [2](https://github.com/romefrontend/rome/blob/main/internal/core/server/fs/MemoryFileSystem.ts#L631), where the second place lacks `manifest` & `path`. We can discuss it further if we decide to move forward.

If we decide to close this PR, I guess I'll wait for the next available `good first issue` 🤓

**Comment by sebmck** [comment]
Yeah I agree we can do better to clarify, I think we should go in the direction I mentioned. #1 is where the changes would need to be added. #2 handles diagnostics that are "thrown". ie. code can do `throw DiagnosticsError(diagnostics);` which is caught by `catchDiagnostics`. We do this so we can add and propagate diagnostics from absolutely anywhere.

File: internal/core/server/fs/MemoryFileSystem.ts [hunk 1]
```diff
@@ -651,11 +652,27 @@ export default class MemoryFileSystem {
...
-		if (normalizedDiagnostics.length > 0) {
+		if (rawDiagnostics.length > 0) {
+			const normalizedDiagnostics = rawDiagnostics.map((diag) => ({
+				...diag,
+				description: {
+					...diag.description,
+					advice: [
+						...diag.description.advice,
+						{
+							type: (<"log">"log"),
```
Comment (barakyosi) [inline internal/core/server/fs/MemoryFileSystem.ts:667]: Not a TS expert 😬 - I got this error without casting here:
```
Argument of type '{ description: { advice: (DiagnosticAdviceLog | DiagnosticAdviceList | DiagnosticAdviceInspect | DiagnosticAdviceCode | ... 6 more ... | { ...; })[]; category: DiagnosticCategory; message: StaticMarkup; }; ... 5 more ...; tags?: DiagnosticTags | undefined; }[]' is not assignable to parameter of type 'Diagnostics'.
  Type '{ description: { advice: (DiagnosticAdviceLog | DiagnosticAdviceList | DiagnosticAdviceInspect | DiagnosticAdviceCode | ... 6 more ... | { ...; })[]; category: DiagnosticCategory; message: StaticMarkup; }; ... 5 more ...; tags?: DiagnosticTags | undefined; }' is not assignable to type 'Diagnostic'.
    The types of 'description.advice' are incompatible between these types.
      Type '(DiagnosticAdviceLog | DiagnosticAdviceList | DiagnosticAdviceInspect | DiagnosticAdviceCode | ... 6 more ... | { ...; })[]' is not assignable to type 'DiagnosticAdvice'.
        Type 'DiagnosticAdviceLog | DiagnosticAdviceList | DiagnosticAdviceInspect | DiagnosticAdviceCode | ... 6 more ... | { ...; }' is not assignable to type 'DiagnosticAdviceItem'.
          Type '{ type: string; category: "info"; text: StaticMarkup; }' is not assignable to type 'DiagnosticAdviceItem'.
            Type '{ type: string; category: "info"; text: StaticMarkup; }' is not assignable to type 'DiagnosticAdviceLog'.
              Types of property 'type' are incompatible.
                Type 'string' is not assignable to type '"log"'.ts(2345)
```

Would be happy for any advice :)

Comment (sebmck) [reply internal/core/server/fs/MemoryFileSystem.ts:667]: I think TypeScript is getting confused with the types and is widening them. You can add a type annotation of `Diagnostics` to `normalizedDiagnostics` and it _should_ go away.

**Comment by sebmck** [comment]
Perfect, thank you! Really happy with this solution and think it will help a lot of people.

**Pull request merged**